File: Find_maxpeak.m
function Fre_Sorted = Find_maxpeak(Freqs, ~, Amps)
% Find_maxpeak 根据幅值大小对频率峰值进行降序排序
% 
% 输入:
%   Freqs : 候选峰值的频率位置 (对应代码中的 S_rls)
%   ~     : 占位符 (对应代码中重复传入的 S_rls，此处忽略)
%   Amps  : 候选峰值的幅值 (对应代码中的 S_rls_amp)
%
% 输出:
%   Fre_Sorted : 按幅值降序排列后的频率数组
%                Fre_Sorted(1) 即为最大峰对应的频率

    % 1. 鲁棒性检查：如果输入为空，直接返回空
    if isempty(Freqs) || isempty(Amps)
        Fre_Sorted = [];
        return;
    end

    % 2. 确保输入为列向量，防止维度不匹配
    Freqs = Freqs(:);
    Amps  = Amps(:);

    % 3. 核心逻辑：对幅值进行降序排序 (Descend)
    % val 没用到，我们只需要排序后的索引 idx
    [~, idx] = sort(Amps, 'descend');

    % 4. 根据排序索引重新排列频率数组
    Fre_Sorted = Freqs(idx);

end

File: python/src/ppg_hr/core/find_maxpeak.py
"""Sort candidate peak frequencies by descending amplitude (port of ``Find_maxpeak.m``)."""

from __future__ import annotations

import numpy as np

__all__ = ["find_maxpeak"]


def find_maxpeak(
    freqs: np.ndarray, _placeholder, amps: np.ndarray
) -> np.ndarray:
    """Return ``freqs`` reordered by descending ``amps``.

    The middle argument is kept for signature parity with the MATLAB caller
    (which passes the same frequency vector twice).
    """
    f = np.atleast_1d(np.asarray(freqs)).ravel()
    a = np.atleast_1d(np.asarray(amps)).ravel()
    if f.size == 0 or a.size == 0:
        return np.array([])
    # MATLAB ``sort(..., 'descend')`` is stable; numpy.argsort default is not,
    # but stability matters only on amplitude ties (rare in float64 spectra).
    idx = np.argsort(-a, kind="stable")
    return f[idx]
